### PR TITLE
Do not snap trimesh vertices to pixel grid

### LINKED ISSUE #806: Trimesh rasterization seems slightly off
Please edit the template below, attempting to fill in as much info as possible (it's all optional, but the more you fill in, the faster we can help). Thanks!

When pasting code, tracebacks, or console output, please enclose in triple backticks (https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code).

## Description

While playing around with Canvas.trimesh and rasterizing simple geomtery it seems that the rule described in the documentation ("a pixel (bin) is treated as belonging to a given triangle if its center falls either inside that triangle or along its top or left edge") is not adhered to, i.e., some pixels that do not belong to a triangle get a value, and the other way around.

rasterized_circle

## How to reproduce

```
import pandas as pd
import shapely.geometry
from scipy.spatial import Delaunay
import datashader as ds
import matplotlib.pyplot as plt


def triangulate(vertices, x="x", y="y"):
    triang = Delaunay(vertices[[x,y]].values)
    return pd.DataFrame(triang.simplices, columns=['v0', 'v1', 'v2'])


shape = shapely.geometry.Point(0, 0).buffer(1)
vertices = pd.DataFrame(shape.exterior.coords, columns=('x', 'y'))
vertices['z'] = 1.0

simplices = triangulate(vertices)

canvas = ds.Canvas(plot_width=10, plot_height=10,
                x_range=(-1, 1), y_range=(-1, 1))

canvas.trimesh(vertices, simplices, agg=ds.any()).plot(alpha=0.5)
plt.plot(vertices.x, vertices.y)

```

### PR DESCRIPTION
**Pull request by ianthomas23**
Closes #806.

Previously `trimesh` snapped each vertex to the nearest pixel center before deciding which pixels were inside each triangle. This worked mostly OK except that it never rendered pixels in the right-hand column or bottom row, and small triangles (just a few pixels wide/high) rendered badly. This fix does not snap vertices to pixel centers to avoid these problems.

Here are before and after images for the three test cases in issue #806, blue lines show the boundary passed to `trimesh` and the last example should completely fill the agg:
before1 after1

before2after2

before3after3

Most of the work here is changing and checking all of the trimesh tests. I have passed a human eyeball over all of the integration test results (`test_pandas` and `test_dask`), and had to rewrite all of the unit tests (`test_glyphs`) as the internal `draw_triangle` functions now take floating-point vertex locations (unsnapped).

### CODE REVIEW

File: datashader/glyphs/trimesh.py [hunk 1]
```diff
@@ -214,10 +227,10 @@ def map_onto_pixel(vt, bounds, x, y):
 
         Points falling on upper bound are mapped into previous bin.
```
Comment (jbednar) [inline datashader/glyphs/trimesh.py:228]: Is this statement still accurate?

**Review by jbednar — APPROVED** [review]
Looks good; thanks so much! Happy for you to merge it after addressing one comment about the docstring.

File: datashader/glyphs/trimesh.py (hunk 1 above)
Comment (jbednar) [reply datashader/glyphs/trimesh.py:228]: (Maybe that same point should be made still, but elsewhere, since it's no longer enforced here?)

Comment (ianthomas23) [reply datashader/glyphs/trimesh.py]: No, it is no longer accurate so I have removed it.

**Comment by ianthomas23** [comment]
I have made another small change so that subpixel rendering (whole triangle within a single pixel) remains the same.

I am as happy with this as I can be. Longer term I think it will be necessary to check the results are consistent with `canvas.polygons` plus also that we are correctly dealing with special situations like long thin axis-aligned meshes (e.g. 1 or 2 pixels in y dimension, many pixels in x dimension) as they are not handled particularly well, both before and after this PR.

**Pull request merged**